src/System.Private.ServiceModel/src/System/ServiceModel/Channels/MessageContent.cs

```diff
@@ -0,0 +1,266 @@
+    {
+        private Task _encodingTask;
+
+        public StreamedMessageContent(Message message, MessageEncoder messageEncoder) : base(message, messageEncoder)
+        {
+        }
+
+
+        protected override Task<Stream> CreateContentReadStreamAsync()
+        {
+            _stream = new ProducerConsumerStream();
+            _encodingTask = _messageEncoder.WriteMessageAsync(_message, _stream);
+            return Task.FromResult(_stream);
+        }
+
+        protected override async Task SerializeToStreamAsync(Stream stream, TransportContext context)
+        {
+            _stream = new ProducerConsumerStream();
+            await _messageEncoder.WriteMessageAsync(_message, _stream);
+            await _stream.CopyToAsync(stream, 8192);
+        }
+
+        protected override bool TryComputeLength(out long length)
+        {
+            length = -1;
+            return false;
+        }
+
+    }
+
+    internal class BufferedMessageContent : MessageContent
+    {
+        private bool _disposed;
+        private bool _messageEncoded;
+        private ArraySegment<byte> _buffer;
+        private BufferManager _bufferManager;
+        protected long? _contentLength;
+
+        public BufferedMessageContent(Message message, MessageEncoder messageEncoder, BufferManager bufferManager) : base(message, messageEncoder)
+        {
+            _bufferManager = bufferManager;
+            _messageEncoded = false;
+        }
+
+        protected override Task<Stream> CreateContentReadStreamAsync()
+        {
+            EnsureMessageEncoded();
+            _stream = new MemoryStream(_buffer.Array, _buffer.Offset, _buffer.Count, false, true);
+            return Task.FromResult(_stream);
+        }
+
+        private void EnsureMessageEncoded()
+        {
+            if (!_messageEncoded)
+            {
+                _buffer = _messageEncoder.WriteMessage(_message, int.MaxValue, _bufferManager);
+                _contentLength = _buffer.Count;
+                _messageEncoded = true;
+            }
+        }
+
+        protected override async Task SerializeToStreamAsync(Stream stream, TransportContext context)
+        {
+            EnsureMessageEncoded();
+            await stream.WriteAsync(_buffer.Array, _buffer.Offset, _buffer.Count);
+        }
+
+        protected override bool TryComputeLength(out long length)
+        {
+            EnsureMessageEncoded();
+            if (_contentLength.HasValue)
+            {
+                length = (long)_contentLength;
+                return true;
+            }
+            else
+            {
+                length = 0;
+                return false;
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && !_disposed)
+            {
+                _disposed = true;
+                if(_buffer.Array != null)
+                {
+                    _bufferManager.ReturnBuffer(_buffer.Array);
```

Comment:
`_buffer = default(ArraySegment<byte>)` 
It would be good if the buffer manager zero'd out the buffer before handing it out.